### PR TITLE
HTTSTB: header processing, see #88

### PR DESCRIPTION
**Pull request by equalsJeffH**
Here's a mostly-complete initial attempt at addressing issue #88. Note that it does not address the "Connection" header and proxies question(s) being discussed on the list (see <https://www.ietf.org/mail-archive/web/unbearable/current/msg01163.html>).

We can either fold the resolution to the "Connection" header and proxies question(s) into this PR before merging, or address that in a separate PR. In any case, there's material here to review now. 

thanks, =JeffH

### CODE REVIEW

File: draft-ietf-tokbind-https-08.xml [hunk 1]
```diff
@@ -297,32 +297,47 @@
...
   Sec-Token-Binding: <base64url-encoded TokenBindingMessage>
         ]]></artwork></figure>
       
-
+      <t>If the server recieves more than one Sec-Token-Binding header field in 
+      an HTTP request, then the server MUST process only the first such header
+      field. If the processed Sec-Token-Binding header field contains more than 
+      one EncodedTokenBindingMessage (e.g., delimited by commas, see Section 3.2 of <xref target="RFC7230"/>),
+      the server MUST ignore the header field. Additionally, the Sec-Token-Binding 
+      header field:
```
Comment (Andrei-Popov) [inline draft-ietf-tokbind-https-08.xml]: We're saying above that only one header is allowed, and only one TB message must be in it. If the client violates these restrictions, then we have a malformed request, and should simply reject this request.

Comment (Andrei-Popov) [inline draft-ietf-tokbind-https-08.xml]: We're saying above that only one header is allowed, and only one TB message must be in it. If the client violates these restrictions, then we have a malformed request, and should simply reject this request.

Comment (Andrei-Popov) [inline draft-ietf-tokbind-https-08.xml]: We're saying above that only one header is allowed, and only one TB message must be in it. If the client violates these restrictions, then we have a malformed request, and should simply reject this request.

Comment (Andrei-Popov) [inline draft-ietf-tokbind-https-08.xml]: We're saying above that only one header is allowed, and only one TB message must be in it. If the client violates these restrictions, then we have a malformed request, and should simply reject this request.

Comment (balfanz) [reply draft-ietf-tokbind-https-08.xml]: Rejecting the request is fine with me.

So is ignoring subsequent header fields.

So is saying that it's ok to submit multiple header fields (just like it's ok to have multiple TokenBindingIDs in a TokenBindingMessage), and just passing them all on to upper layers - maybe that's how the client shows that it possesses multiple keys?

(Maybe I have a slight preference for the latter.)

Andrei - do you feel strongly about this? I'm willing to merge any of those options into the main branch.

Comment (nharper) [reply draft-ietf-tokbind-https-08.xml]: My preference would be to reject a request with multiple Sec-Token-Binding headers.

**Comment by Andrei-Popov** [comment]
I think the spec should be definitive: since we’re saying the client should only send one TB header per request, and only one TB message per header, then the server should be at liberty to enforce this.

If we allow variation, e.g. it’s OK to send multiple TB headers per request, then we need to define when and how one would use this mechanism. And at this point in the process I don’t think we should be adding new mechanisms to TB v 1.0.

Cheers,

Andrei

File: draft-ietf-tokbind-https-08.xml (hunk 1 above)
Comment (b---c) [reply draft-ietf-tokbind-https-08.xml]: +1 to reject a request with multiple Sec-Token-Binding headers

**Comment by balfanz** [comment]
Don't we allow multiple TokenBinding structures within a TokenBindingMessage? Seems to me that the server already has to be able to deal with multiple proofs-of-possession. Allowing multiple header fields is just another way that we might end up with multiple proofs-of-possession - it's not exactly a "new mechanism".

But like I said, I don't feel strongly about it. I'm happy to have the spec say that the server should (must?) reject if there are multiple headers.

**Comment by Andrei-Popov** [comment]
Correct, we do allow multiple bindings within a TB message (and to me that's an existing mechanism:).
Supporting the same thing via multiple header fields requires a code change, and so far I see no reason to maintain two ways (or mechanisms:) to do the same thing.

**Comment by equalsJeffH** [comment]
pls see reply on-list: 

https://www.ietf.org/mail-archive/web/unbearable/current/msg01193.html

**Comment by equalsJeffH** [comment]
this comment thread has been taken to unbearable@ietf.org:

https://www.ietf.org/mail-archive/web/unbearable/current/msg01193.html
https://www.ietf.org/mail-archive/web/unbearable/current/msg01194.html
https://www.ietf.org/mail-archive/web/unbearable/current/msg01195.html
https://www.ietf.org/mail-archive/web/unbearable/current/msg01196.html

File: draft-ietf-tokbind-https-08.xml [hunk 2]
```diff
@@ -193,7 +193,7 @@
       </address>
     </author>
 
-    <date year="2016" />
+    <date year="2016" month="Feb" />
```
Comment (Andrei-Popov) [inline draft-ietf-tokbind-https-08.xml]: I think it's better to update the year.

Comment (Andrei-Popov) [reply draft-ietf-tokbind-https-08.xml]: Otherwise, the change looks good to me.

**Comment by equalsJeffH** [comment]
Ok, so let's not say anything about the `Connection` header and see what happens.  ship it.

**Comment by ve7jtb** [comment]
OK then can we publish the updated drafts.

John B.

**Comment by Andrei-Popov** [comment]
Uploaded to IETF.

Cheers,

Andrei

**Pull request merged**